File: go/rps.go
package main

import (
	"fmt"
	"math/rand"
	"strings"
	"time"
)

func main() {
	// Seed random number generator. https://gobyexample.com/random-numbers
	rand.Seed(time.Now().UnixNano())

	var choice string
	choices := map[int]int{
		// choice : beats (0 -> Rock, 1 -> Paper, 2 -> Scissors)
		1: 0,
		0: 2,
		2: 1,
	}

loop:
	for {
		computerMove := rand.Intn(3)
		humanMove := -1

		// get player move and lowercase
		fmt.Print("Your move: ")
		fmt.Scanf("%s\n", &choice)
		choice = strings.ToLower(choice)

		if choice == "rock" {
			humanMove = 0
		} else if choice == "paper" {
			humanMove = 1
		} else if choice == "scissors" {
			humanMove = 2
		}

		if humanMove != -1 {
			// calculate winner
			switch humanMove {
			case computerMove:
				fmt.Println("Tie, Replay!")
			case choices[computerMove]:
				fmt.Println("Computer Wins!")
				break loop
			default:
				fmt.Println("You Win!")
				break loop
			}
		}
	}
}

File: c++/rps.cpp
#include <ctime>
#include <iostream>
#include <string>
#include <algorithm>
#include <map>

using namespace std;

int main() {
  // map possible choices to the moves they beat
  map<string, string> choices;
  choices["rock"] = "scissors";
  choices["paper"] = "rock";
  choices["scissors"] = "paper";

  string human_move, computer_move;
  int size = choices.size();

  while ( true ) {
    // seed random number generator
    srand(time(NULL));

    // calculate computer move
    map<string, string>::iterator it = choices.begin();
    advance(it, rand() % size);
    computer_move = it->first;

    // get player input, must be valid
    cout << "Your Move: ";
    getline(cin, human_move);

    // lowercase input
    // source: https://notfaq.wordpress.com/2007/08/04/cc-convert-string-to-upperlower-case/ 
    transform(human_move.begin(), human_move.end(), human_move.begin(), ::tolower);

    // validate user input
    if ( choices.find(human_move) == choices.end() ) {
      // invalid move 
      continue;
    }

    // calculate winner
    if ( computer_move == human_move ) {
      cout << "Tie, Replay!\n\n";
      human_move = -1;
    } else if ( choices[human_move] == computer_move ) {
      cout << "You Win!\n\n";
      break;
    } else if ( choices[computer_move] == human_move ) {
      cout << "Computer Wins!\n\n";
      break;
    }
  }

  return 0;
}
